proposals/0000-single-pattern-checks.md

```diff
@@ -0,0 +1,106 @@
+# Single pattern checks
+
+* Proposal: [HXP-NNNN](NNNN-single-pattern-checks.md)
+* Author: [Dmitrii Maganov](https://github.com/vonagam)
+
+## Introduction
+
+New syntax for single pattern checks.
+
+## Motivation
+
+Current single pattern check - `match` method on an enum value - is limited in functionality compared to usual pattern matching. No captures or guards and only a single enum value as a root value.
+
+Proposed syntax will allow to use full arsenal of matcher for three common scenarios.
+
+## Detailed design
+
+All of single pattern checks allias such switch:
+
+```haxe
+switch ($value) {
+  case $pattern if ($guard): $ifBody;
+  case _: $elseBody;
+}
+```
+
+### Pattern test with `is case`
+
+```
+$value is case $pattern [where ($guard)]
+```
+
+With `$ifBody` being `true` and `$elseBody` - `false`.
+
+```haxe
+function isSome(option: Option<Int>): Bool {
+  return option is case Some(_);
+}
+
+function isDivisibleSome(option: Option<Int>, divisor: Int): Bool {
+  return option is case Some(int) where (int % divisor == 0);
+}
+```
+
+### Pattern condition with `if case`
+
+```
+if case ($value, $pattern[, $guard]) $ifBody [else $elseBody]
+```
+
+```haxe
+function ifSome(option: Option<Int>, then: (int: Int) -> Void): Void {
+  if case (option, Some(int)) {
+    then(int);
+  }
+}
+
+function ifDivisibleSome(option: Option<Int>, divisor: Int, then: (int: Int) -> Void, or: () -> Void): Void {
+  if case (option, Some(int), int % divisor == 0) {
+    then(int);
+  } else {
+    or();
+  }
+}
+```
+
+### Pattern assertion with `cast case`
```

Comment:
This one seems really obscure to me. I get it with the examples, but code itself is not self-explanatory imo.

The thing bothering me the most here, I think, is that `$ifBody` is not inside its own code block but has its own scope (capture variables), etc.

Reply:
Last one is the aim of it - to reduce nesting for common scenario of throwing error or exiting early.

Reply:
Additional counter for reference:
Search for `(_|None) ->[\s\n]+(true|false)` - 1128 matches.
So match where you want to get a boolean is not more popular than one where you handle input of specific patterns and anything else is an early exit (usually in an error form).

Reply:
`$ifBody` is inside and what it does should be clear now.